### PR TITLE
Allow being able to add extendable task items for the task list

### LINKED ISSUE #18: Improve task list creation
This would be an improvement upon the initial task list generation -> https://github.com/woocommerce/woocommerce-admin-test-helper/pull/17

We currently only set a number of tasks to be generated, and it auto generates them. Using `Math.random()` for the completed and dismissible variables.

This ticket would be to improve this, allowing a user to create task items and being able to set the title, completed, visible, additionalInfo, time, isDimissable, and type. This could all be saved in the an option and set in a similar manner as what we are doing with `window.wcAdminTestHelperConfig`.

### PR DESCRIPTION
**Pull request by louwie17**
Fixes: #18 

~~This is an initial PR for being able to add extendable task items for the task list.
Currently I only have it set to a specific number, and it will auto generate the number of task items, it would be nice to change it so we can add all the task lists manually, having control over (title, dismisable, completed, etc...) (created an issue for it -> https://github.com/woocommerce/woocommerce-admin-test-helper/issues/18).~~

There are two main changes in this PR:
- I updated the script enqueue logic to only load the app on the **Tools > WCA Test Helper** page and adding a separate bundle used for the filters that gets enqueued on every admin page.
- I created a extended task list tab, with a list of test helper tasks, someone can also create new tasks with a couple different options as shown in the GIF below. Note: The task list type can either be `setup` or `extension`, if set to `setup` it will show in the main task list, if set to `extension` it is shown as an extension task list (also shown in the GIF).

admin-test-helper-task-list

### Testing instructions

- Activate the WC Admin Test Helper plugin with this branch loaded
- Go to **Tools > WCA Test Helper > Extended Tasks**
- Create a couple tasks, one of each type
- Navigate to **WooCommerce > Home**
- Newly created tasks should be displayed (if task list is enabled) you should be able to click on them and complete them.
- Navigate back to **Tools > WCA Test Helper > Extended Tasks** and edit and or delete a task
- Go back to **WooCommerce > Home** the tasks should not be displayed if deleted or correctly updated if edited (can only edit is completed).

### CODE REVIEW

File: woocommerce-admin-test-helper.php [hunk 1]
```diff
@@ -54,7 +58,7 @@ function add_extension_register_script() {
 	wp_enqueue_style( 'woocommerce-admin-test-helper' );
 }
 
-add_action( 'admin_enqueue_scripts', 'add_extension_register_script' );
+add_action( 'admin_enqueue_scripts', 'add_wc_admin_helper_extension_register_script' );
```
Comment (louwie17) [inline woocommerce-admin-test-helper.php:61]: Update this, so it does not conflict with any other auto generated plugins

File: src/tasks/add-tasks.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { addFilter } from '@wordpress/hooks';
+
+const TaskComp = () => {
```
Comment (becdetat) [inline src/tasks/add-tasks.js:3]: what is a `TaskComp`? Could this be inlined in `generateTasks`?

**Comment by becdetat** [comment]
I'm finding the interface very confusing, it isn't clear what I'm doing with tasks. Also what is `add-tasks.js` doing? It would appear to create 5 tasks but I can't see them in the tasks list.

File: src/tasks/add-tasks.js (hunk 2 above)
Comment (louwie17) [reply src/tasks/add-tasks.js:3]: I should rename this, it is the container that is shown when clicking on a task.

**Comment by louwie17** [comment]
> I'm finding the interface very confusing, it isn't clear what I'm doing with tasks. Also what is `add-tasks.js` doing? It would appear to create 5 tasks but I can't see them in the tasks list.

Maybe I will work on this a little more then, going to **Tools > WCA Test Helper > Tasks** allows you to set the number of tasks. Then when you load the home screen it should display the extendable task list with the number of tasks.
`add-tasks.js` randomly generates the tasks.
If it doesn't show up for you, I wonder if task list or extendable task list is set to hidden?

This was my current flow:
test-helper-task-list

**Comment by becdetat** [comment]
The task list was showing up, it just wasn't clear what the tool was doing

File: webpack.config.js [hunk 3]
```diff
@@ -1,11 +1,15 @@
-const defaultConfig = require( '@wordpress/scripts/config/webpack.config' );
-const WooCommerceDependencyExtractionWebpackPlugin = require( '@woocommerce/dependency-extraction-webpack-plugin' );
+const defaultConfig = require('@wordpress/scripts/config/webpack.config');
+const WooCommerceDependencyExtractionWebpackPlugin = require('@woocommerce/dependency-extraction-webpack-plugin');
 
 module.exports = {
 	...defaultConfig,
+	entry: {
+		...defaultConfig.entry,
+		'extended-tasks': './src/extended-tasks/include-extended-tasks.js',
```
Comment (louwie17) [inline webpack.config.js:8]: Add a separate entry for the task list filter logic.

File: woocommerce-admin-test-helper.php [hunk 4]
```diff
@@ -15,7 +15,10 @@
 /**
  * Register the JS.
  */
-function add_extension_register_script() {
+function add_wc_admin_helper_extension_register_script( $hook_suffix ) {
+	if ( 'tools_page_woocommerce-admin-test-helper' !== $hook_suffix ) {
+		return;
+	}
```
Comment (louwie17) [inline woocommerce-admin-test-helper.php:21]: Only `enqueue` the test-helper app js files when we are loading the test-helper page.
This way we don't load the JS files on any other wc-admin pages.

File: woocommerce-admin-test-helper.php [hunk 5]
```diff
@@ -54,7 +57,33 @@ function add_extension_register_script() {
...
+	wp_enqueue_script(
+		'woocommerce-admin-test-helper-extended-tasks',
+		$script_url,
+		$script_asset['dependencies'],
+		$script_asset['version'],
+		true
+	);
+}
+
+add_action( 'admin_enqueue_scripts', 'add_extended_task_list_script' );
```
Comment (louwie17) [inline woocommerce-admin-test-helper.php:86]: This loads the task-list filter logic separately on any admin page, although only used on the **WooCommerce > Home** page.
It loads the inline `wcAdminTestHelperConfig` and the `extended-tasks.js` file, bundled separately as shown in the webpack.config

**Comment by louwie17** [comment]
> The task list was showing up, it just wasn't clear what the tool was doing

@becdetat Thanks for the review, I updated the tasks list UI quite a bit, it should be a lot more clear now what the **Extended Tasks** tab is doing and the use case behind it.

**Comment by mattsherman** [comment]
@louwie17 @becdetat This PR has been sitting "Ready to review" for a month. Let's either wrap this up or drop it. 

@becdetat If you are cool with anybody else doing the follow-up review and approving, just let us know. I want to give you a chance to chime in first since you originally requested changes.

**Comment by becdetat** [comment]
@mattsherman I'm fine with someone else doing another review, time is my enemy 😭 

Thanks for making the changes @louwie17

**Comment by mattsherman** [comment]
@louwie17 Is this still valid? Amazingly, it only has a `package-lock.json` conflict. Should the `package-lock.json` changes even be in this PR?

Let's either push to have this reviewed by @woocommerce/mothra or close it out.

**Comment by louwie17** [comment]
Thanks for the ping @mattsherman, this won't be valid anymore given our new REST task list API, I am going to close this.

**Pull request closed without merging**